### PR TITLE
Fixed #25316 -- Fixed using order_by() and values() after annotate().

### PR DESCRIPTION
**Pull request by varunnaganathan**
In reference to ticket https://code.djangoproject.com/ticket/25316 .

### CODE REVIEW

**Comment by timgraham** [comment]
Please try to add tests in `tests/expressions/tests.py` instead of in a new test app. It's better to reuse existing models if possible as adding more models increases the complexity and execution time of the test suite.

**Comment by jarshwah** [comment]
Rather, there's already a `tests/expressions_case/tests.py` that could be used that already has the model you're writing tests against. You should move the tests into that file.

**Comment by varunnaganathan** [comment]
Sorry for that.will keep that in mind.
Added the tests to the tests/expressions/tests.py using the existing models.

**Comment by timgraham** [comment]
Could you try to follow the style of the other tests and clean up the code a bit? See the flake8 and isort errors, remove the commented out print statements, etc.

File: tests/expressions/tests.py [hunk 1]
```diff
@@ -883,3 +883,37 @@ def test_aggregates(self):
...
+                discount=Case(
+                    When(salary=10,then=Value(5)),
+                    When(salary=20,then=Value(10)),
+                    When(salary=30,then=Value(15)),
+                    default=Value(0),
+                    output_field=IntegerField()
+                    ),
+                ).order_by('id').values('id')
+        #print (a)
+        self.assertEqual(str(a),str([{u'id': 1}, {u'id': 2}, {u'id': 3}]))
```
Comment (timgraham) [inline tests/expressions/tests.py]: u'' string prefixes aren't needed. We use `from __future__ import unicode_literals`.

File: tests/expressions/tests.py [hunk 2]
```diff
@@ -883,3 +883,37 @@ def test_aggregates(self):
...
+    """checking if it works with annotated field included in the order_by"""
```
Comment (timgraham) [inline tests/expressions/tests.py]: Docstrings go inside each method.

**Comment by varunnaganathan** [comment]
Hi,I made all the changes you suggested.
I ran flake8 on my command line ,but no errors came up.The line it is showing the error was not even part of any changes I made.Further I'm unable to find the error it is suggesting.The error shown is:
./django/db/models/expressions.py:761:31: E231 missing whitespace after ',' 
Other than that,all other tests pass.I changed the coding style in accordance with other tests.Will keep that in mind in the future.

File: tests/expressions_case/tests.py [hunk 3]
```diff
@@ -1243,3 +1245,40 @@ def test_conditional_aggregation_example(self):
...
+        """
+        checking if it works with annotated field not included in the order_by
+        """
+        a = Client.objects.annotate(
+            discount=Case(
+                When(account_type=Client.GOLD, then=Value('5%')),
+                When(account_type=Client.PLATINUM, then=Value('10%')),
+                default=Value('0%'),
+                output_field=CharField(),
+                ),
```
Comment (timgraham) [inline tests/expressions_case/tests.py]: deindent this line and the next line by 4 spaces

File: tests/expressions_case/tests.py [hunk 4]
```diff
@@ -1243,3 +1245,40 @@ def test_conditional_aggregation_example(self):
...
+            ).order_by('id').values('id')
+        self.assertEqual(str(a), str([{'id': 1}, {'id': 2}, {'id': 3}]))
```
Comment (timgraham) [inline tests/expressions_case/tests.py]: `(list(a), [...])` should be simpler

File: django/db/models/expressions.py [hunk 5]
```diff
@@ -719,7 +719,8 @@ def get_source_fields(self):
     def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
         c = self.copy()
         c.is_summary = summarize
-        c.condition = c.condition.resolve_expression(query, allow_joins, reuse, summarize, False)
+        if hasattr(c.condition,"resolve_expression"):
```
Comment (timgraham) [inline django/db/models/expressions.py]: The flake8 error is due to missing a space after the comma here. Please use single quotes instead of double.

**Comment by timgraham** [comment]
Could you please squash your commits when updating and use the message from the pull request's title?

**Comment by timgraham** [comment]
buildbot, add to whitelist.

File: tests/expressions_case/tests.py (hunk 4 above)
Comment (varunnaganathan) [inline tests/expressions_case/tests.py]: Didn't understand what u mean by this?

Comment (timgraham) [inline tests/expressions_case/tests.py]: cast to list instead of str

**Comment by varunnaganathan** [comment]
Haven't squashed commits before.Please give me a while to figure it out.

**Comment by varunnaganathan** [comment]
Squashed the commits .Please check.

File: .idea/django.iml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (timgraham) [inline .idea/django.iml]: Oops, you probably want to add `.idea` to `.git/info/exclude` or `core.excludesFile` per https://git-scm.com/docs/gitignore.

**Comment by varunnaganathan** [comment]
Hi,Fixed the .idea directory.I'm sorry I'm taking time to getting adjusted to the git workflow.Will take care of the following in the future.
Please review

File: tests/expressions_case/tests.py [hunk 7]
```diff
@@ -1243,3 +1243,31 @@ def test_conditional_aggregation_example(self):
             ),
             {'regular': 2, 'gold': 1, 'platinum': 3}
         )
+
+    def test_annotation_with_values_and_when(self):
```
Comment (timgraham) [inline tests/expressions_case/tests.py]: I think we could include a more descriptive test name instead of a docstring, e.g. `test_annotate_order_by_values_annotation_not_in_order_by`. Also, these tests are in `CaseDocumentationExamples` which is a bit confusing since these aren't documentation examples. Any reason to use something like [the simpler test](https://code.djangoproject.com/attachment/ticket/25316/25312-test.diff) that I attached to the ticket?

**Comment by timgraham** [comment]
Josh, any opinion about backporting this fix?

File: tests/expressions_case/tests.py (hunk 7 above)
Comment (varunnaganathan) [inline tests/expressions_case/tests.py]: Which tests are in the CaseDocumentationExamples ? Also the tests you attached are pretty much similar.I however attached two sets of tests one without and one with the annotated fields in the orderby to ensure better test coverage.

Comment (timgraham) [inline tests/expressions_case/tests.py]: You can grep the docs to find the examples. At least some are in `docs/ref/models/conditional-expressions.txt`. I don't mind adding more tests, but it's a bit confusing to add them as part of a commit which fixes a different bug, so for clarity please make it a separate commit at least. Thanks!

Comment (varunnaganathan) [inline tests/expressions_case/tests.py]: Ya ok,so then I'll remove the extra test.As far as the simplicity of tests us concerned,I feel the tests are included and the ones you included in the diff are pretty much the same.Do you want me to still remove my tests and use your tests?And about the existing tests in the docs,do you want me to remove them or add those to the expressions_case/tests.py ?

Comment (timgraham) [inline tests/expressions_case/tests.py]: I think you can apply my test patch and that should be sufficient.

**Comment by jarshwah** [comment]
I think it's fine to backport considering it's a crashing bug with a valid use case.

**Comment by timgraham** [comment]
Would like to include this in today's release, so updated in https://github.com/django/django/pull/5916.
